src/actions/toggleNote.ts

```diff
@@ -1,37 +1,102 @@
+import createId from '../util/createId'
 import head from '../util/head'
 import reducerFlow from '../util/reducerFlow'
+import resolvePathReference from '../util/resolvePathReference'
 
 /** Toggle the caret between the cursor and its note. Set the selection to the end. If the note is empty, delete it. */
 const toggleNote = (state: State): State => {
   const thoughtId = head(state.cursor!)
   const hasNote = findDescendant(state, thoughtId, '=note')
-  const noteChildren = hasNote ? getChildren(state, hasNote) : []
-  const offset = noteChildren.length ? noteChildren[0].value.length : 0
 
-  return reducerFlow([
-    // create an empty note if it doesn't exist
-    !hasNote
-      ? setDescendant({
-          path: state.cursor!,
-          values: ['=note', ''],
+  // Check if there's a path reference
+  const pathReference = resolvePathReference(state, thoughtId, hasNote)
+
+  // Calculate offset and handle target thought creation if needed
+  let offset = 0
+  let newState = state
+
+  if (pathReference) {
+    // Start with the targetId from pathReference (could be undefined if target doesn't exist)
+    let targetId = pathReference.targetId
+
+    // Create the missing target thought if it doesn't exist
+    if (!pathReference.targetExists) {
+      // Calculate appropriate rank for the new thought
+      const pathRank = getNextRank(state, thoughtId)
+
+      // Create the target thought
+      newState = createThought(state, {
+        id: createId(),
+        path: [thoughtId],
+        value: pathReference.pathValue,
+        rank: pathRank,
+      })
+
+      // Find the newly created thought: this should not be null after creation
+      const newTargetId = findDescendant(newState, thoughtId, pathReference.pathValue)
+      if (newTargetId) {
+        targetId = newTargetId
+      }
+    }
+
+    if (targetId) {
+      // Check if the target thought has children
+      const targetChildren = getChildren(newState, targetId)
+
+      // If the target has no children, create an empty child
+      if (targetChildren.length === 0) {
+        // Calculate appropriate rank for the child thought
+        const childRank = getNextRank(newState, targetId)
+
+        // Create an empty child for the target path
+        newState = createThought(newState, {
+          path: [targetId],
+          value: '',
+          rank: childRank,
         })
-      : // delete an empty note that already exists
-        state.noteFocus && !attribute(state, thoughtId, '=note')
-        ? deleteAttribute({ path: state.cursor!, value: '=note' })
+
+        offset = 0
+      } else {
+        // If target already has children, set offset to the length of the first child's value
+        offset = targetChildren[0].value.length
+      }
+    }
+  } else if (hasNote) {
+    const noteChildren = getChildren(newState, hasNote)
+    offset = noteChildren.length ? noteChildren[0].value.length : 0
+  }
+
+  const finalState = reducerFlow([
```

Comment:
There should not be multiple state variables (`state`, `newState`, `finalState`). Take advantage of `reducerFlow` to compose chains of reducers. There are plenty of examples of this, including nested reducers, throughout the codebase.